File: tensorrt_llm/runtime/kv_cache_manager_v2/_page.py
# SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from collections.abc import Sequence
from dataclasses import dataclass, field
from typing import TYPE_CHECKING, NamedTuple, cast

from . import rawref
from ._block_radix_tree import Block
from ._common import (
    BAD_PAGE_INDEX,
    DEFAULT_BEAM_INDEX,
    GPU_LEVEL,
    NDEBUG,
    BeamIndex,
    BlockOrdinal,
    CacheLevel,
    PageIndex,
    PageStatus,
    Priority,
    TokenIdExt,
)

if TYPE_CHECKING:
    from ._core._kv_cache import _KVCache
    from ._storage_manager import StorageManager


from ._eviction_controller import NodeRef
from ._exceptions import LogicError
from ._life_cycle_registry import LifeCycleId
from ._storage._core import Slot
from ._utils import (
    CachedCudaEvent,
    assert_critical,
    filled_list,
    get_uniform_attribute,
    merge_events,
    partition,
    stream_wait_events,
    unwrap_rawref,
)

ReferenceType = rawref.ReferenceType


# We will have a huge amount of pages for large storage capacity.
# So we prefer inheritance over composition to save some memory.
@dataclass(slots=True)
class Page(Slot):
    _manager: ReferenceType["StorageManager"]
    life_cycle: LifeCycleId
    cache_level: CacheLevel
    _priority: Priority
    # _holder is either None or a valid rawref.
    _holder: ReferenceType["_PageHolder"] | None
    node_ref: NodeRef | None

    def __del__(self) -> None:
        if not NDEBUG:
            assert_critical(self.status == PageStatus.DROPPABLE and not self.scheduled_for_eviction)
        if self.has_valid_slot:
            self.manager.release_slot(self.life_cycle, self.cache_level, self)

    @property
    def manager(self) -> "StorageManager":
        return unwrap_rawref(self._manager)

    @property
    def priority(self) -> Priority:
        return self._priority

    # prevent dropping
    def hold(self) -> "_PageHolder":
        if self._holder is not None:
            return unwrap_rawref(self._holder)
        holder = _PageHolder(self)
        self._holder = rawref.ref(holder)
        controller = self.manager
        if self.scheduled_for_eviction and not controller.is_evictable(self):
            controller.exclude_from_eviction(self)
            assert not self.scheduled_for_eviction
        return holder

    # Prevent eviction. You need to migrate the page to GPU later.
    def lock(
        self,
        kv_cache: "_KVCache",
        beam_index: BeamIndex,
        ordinal: BlockOrdinal,
        life_cycle: LifeCycleId,
        skip_wait: bool = False,
    ) -> "_SharedPageLock":
        "If skip wait, you are responsible for making the page ready in kv_cache.cuda_stream."
        return self.hold().lock(kv_cache, beam_index, ordinal, life_cycle, skip_wait)

    @property
    def status(self) -> PageStatus:
        if self._holder is None:
            return PageStatus.DROPPABLE
        lock_ref = unwrap_rawref(self._holder)._lock
        if lock_ref is None:
            return PageStatus.HELD
        assert unwrap_rawref(lock_ref) is not None
        return PageStatus.LOCKED

    @property
    def scheduled_for_eviction(self) -> bool:
        return self.node_ref is not None

    def is_committed(self) -> bool:
        raise LogicError("Unexpected call to this implementation.")


@dataclass(slots=True)
class UncommittedPage(Page):
    # @TODO: consider move this to _PageHolder
    kv_cache: rawref.ref["_KVCache"]
    ordinal: BlockOrdinal
    beam_index: BeamIndex

    tokens: list[TokenIdExt] = field(default_factory=list)

    def is_committed(self) -> bool:
        return False

    def __init__(
        self,
        kv_cache: "_KVCache",
        ordinal: BlockOrdinal,
        life_cycle: LifeCycleId,
        cache_level: CacheLevel,
        slot: Slot,
        beam_index: BeamIndex = DEFAULT_BEAM_INDEX,
    ):
        self.kv_cache = rawref.ref(kv_cache)
        self.ordinal = ordinal
        self.beam_index = beam_index
        manager = kv_cache.manager
        priority = kv_cache._get_priority(ordinal, manager._life_cycles.get_life_cycle(life_cycle))
        Page.__init__(
            self,
            None,
            CachedCudaEvent.NULL,
            rawref.ref(manager._storage),
            life_cycle,
            cache_level,
            priority,
            None,
            None,
        )
        self.set_slot(slot)

    def convert_to_committed(self, block: Block) -> "CommittedPage":
        """
        Moves the slot to a new committed page and add the new page to the block.
        The uncommitted page becomes invalid.
        """
        assert not self.scheduled_for_eviction
        assert block.storage[self.life_cycle] is None
        # If you hit this assertion failure, it's likely because you are using debugpy, which delayed GC
        # for _KVCache._take_uncommitted_page(). Disable breakpoints on exceptions to avoid this issue.
        assert self.status == PageStatus.DROPPABLE, "Release holder/lock first"
        committed_page = CommittedPage(
            self.manager, block, self.life_cycle, self.cache_level, self, self.priority
        )
        self._slot_id = None
        self.ready_event = CachedCudaEvent.NULL
        assert committed_page.has_valid_slot
        block.storage[self.life_cycle] = rawref.ref(committed_page)
        return committed_page

    def __del__(self) -> None:
        def check_page(p: "BlockPage") -> bool:
            return p is None or isinstance(p.page, CommittedPage)

        if not NDEBUG:
            assert_critical(
                len(unwrap_rawref(self.kv_cache)._blocks) <= self.ordinal
                or check_page(
                    unwrap_rawref(self.kv_cache)
                    ._blocks[self.ordinal]
                    .pages[self.beam_index][self.life_cycle]
                )
            )
        Page.__del__(self)


@dataclass(slots=True)
class CommittedPage(Page):
    block: rawref.ref["Block"]
    __rawref__: rawref.ref["CommittedPage"]

    def is_committed(self) -> bool:
        return True

    def __init__(
        self,
        storage: "StorageManager",
        block: Block,
        life_cycle: LifeCycleId,
        cache_level: CacheLevel,
        slot: Slot,
        priority: Priority,
    ):
        self.block = rawref.ref(block)
        self.__rawref__ = rawref.NULL
        Page.__init__(
            self,
            None,
            CachedCudaEvent.NULL,
            rawref.ref(storage),
            life_cycle,
            cache_level,
            priority,
            None,
            None,
        )
        self.set_slot(slot)

    def __del__(self) -> None:
        block = self.block()
        # block may be None when rebase happens, i.e. another block with the same key is committed,
        # replacing it, but the page is still used by a _KVCache.
        if block is not None:
            block.unset_page(
                self.life_cycle,
                self.manager._life_cycles.get_life_cycle(self.life_cycle),
            )
        Page.__del__(self)
        self.__rawref__.invalidate()


@dataclass(slots=True)
class _PageHolder:
    "Prevents pages from being dropped."

    page: Page
    _lock: rawref.ref["_UniqPageLock"] | None = None
    __rawref__: rawref.ref["_PageHolder"] = field(default_factory=lambda: rawref.NULL)

    def __init__(self, page: Page) -> None:
        self.page = page
        self._lock = None
        self.__rawref__ = rawref.NULL

    def __del__(self) -> None:
        if not NDEBUG:
            assert_critical(self._lock is None)
        page = self.page
        page._holder = None
        # If a held page was in last level cache, it was not scheduled for eviction.
        if page.is_committed():
            page = cast(CommittedPage, page)
            if not page.scheduled_for_eviction:
                page.manager.schedule_for_eviction(page)
            block = page.block()
            if block is None or block.is_orphan:
                page.manager.exclude_from_eviction(page)
        elif page.scheduled_for_eviction:
            page = cast(UncommittedPage, self.page)
            page.manager.exclude_from_eviction(self.page)
        self.__rawref__.invalidate()

    # Prevent eviction. You need to migrate the page to GPU later.
    def lock(
        self,
        kv_cache: "_KVCache",
        beam_index: BeamIndex,
        ordinal: BlockOrdinal,
        life_cycle: LifeCycleId,
        skip_wait: bool = False,
    ) -> "_SharedPageLock":
        if self._lock is None:
            lock = _UniqPageLock(self)
            self._lock = rawref.ref(lock)
        else:
            lock = unwrap_rawref(self._lock)
        if self.page.scheduled_for_eviction:
            manager = self.page.manager
            manager.exclude_from_eviction(self.page)
            assert not self.page.scheduled_for_eviction
        return lock.share(kv_cache, beam_index, ordinal, life_cycle, skip_wait)


@dataclass(slots=True)
class _UniqPageLock:
    "Locks pages to prevent eviction."

    holder: _PageHolder | None
    finish_events: list[CachedCudaEvent]
    __rawref__: rawref.ref["_UniqPageLock"] = field(default_factory=lambda: rawref.NULL)

    def __init__(self, holder: _PageHolder) -> None:
        if holder.page.cache_level != CacheLevel(0):
            raise ValueError("Lock can be applied only on GPU memory pages.")
        self.holder = holder
        self.finish_events = []
        self.__rawref__ = rawref.NULL

    def share(
        self,
        kv_cache: "_KVCache",
        beam_index: BeamIndex,
        ordinal: BlockOrdinal,
        life_cycle: LifeCycleId,
        skip_wait: bool,
    ) -> "_SharedPageLock":
        ret = _SharedPageLock(self, kv_cache, beam_index, ordinal, life_cycle, skip_wait)
        return ret

    @property
    def page(self) -> Page:
        assert self.holder is not None
        return self.holder.page

    def __del__(self) -> None:
        page = self.page
        if not NDEBUG:
            assert_critical(page.cache_level == CacheLevel(0) and not page.scheduled_for_eviction)
        page.ready_event = merge_events(self.finish_events)
        assert self.holder is not None
        self.holder._lock = None
        if False:
            if page.manager.is_evictable(page):
                page.manager.schedule_for_eviction(page)
        else:
            # Optimized code path:
            # delete holder first, so if nobody holds the page elsewhere, it becomes droppable immediately,
            # before we hand it over to eviction controller.
            self.holder = None
            # if it's not droppable, then it means self.holder=None had no impact. We need to schedule it
            # for eviction as usual.
            if page.status != PageStatus.DROPPABLE and page.manager.is_evictable(page):
                page.manager.schedule_for_eviction(page)
        self.__rawref__.invalidate()


class LockOwner(NamedTuple):
    kv_cache: rawref.ref["_KVCache"]
    beam_index: BeamIndex
    ordinal: BlockOrdinal
    life_cycle: LifeCycleId


@dataclass(slots=True, init=False)
class _SharedPageLock:
    _uniq_lock: _UniqPageLock | None
    _user: LockOwner

    @property
    def page(self) -> Page:
        assert self._uniq_lock is not None
        return self._uniq_lock.page

    @property
    def holder(self) -> _PageHolder:
        assert self._uniq_lock is not None
        assert self._uniq_lock.holder is not None
        return self._uniq_lock.holder

    def __hash__(self) -> int:
        return hash(id(self))

    def __eq__(self, other: object) -> bool:
        return self is other

    def __init__(
        self,
        uniq_lock: _UniqPageLock,
        kv_cache: "_KVCache",
        beam_index: BeamIndex,
        ordinal: BlockOrdinal,
        life_cycle: LifeCycleId,
        skip_wait: bool,
    ) -> None:
        self._uniq_lock = uniq_lock
        if not skip_wait:
            self.page.ready_event.wait_in_stream(kv_cache.cuda_stream)
        self._user = LockOwner(rawref.ref(kv_cache), beam_index, ordinal, life_cycle)
        new_index = self._get_page_index()
        old_index = kv_cache._update_page_index(beam_index, ordinal, life_cycle, new_index)
        assert old_index == BAD_PAGE_INDEX

    def __del__(self) -> None:
        if self._uniq_lock is not None:
            self.unlock()

    def unlock(self) -> Page:
        assert self._uniq_lock is not None
        page = self.page
        self._uniq_lock.finish_events.append(unwrap_rawref(self._user.kv_cache).finish_event)
        new_index = BAD_PAGE_INDEX
        old_index = unwrap_rawref(self._user.kv_cache)._update_page_index(
            self._user.beam_index, self._user.ordinal, self._user.life_cycle, new_index
        )
        assert NDEBUG or old_index == self._get_page_index()
        self._uniq_lock = None
        return page

    def _get_page_index(self) -> PageIndex:
        storage = unwrap_rawref(self._user.kv_cache).manager._storage
        num_buffers_per_slot = storage._slot_to_page_indices[self._user.life_cycle]
        return PageIndex(self.page.slot_id * num_buffers_per_slot)


BlockPage = _SharedPageLock | _PageHolder | None


class BatchedLockTarget(NamedTuple):
    page: Page
    beam_index: BeamIndex
    ordinal: BlockOrdinal
    life_cycle: LifeCycleId


def batched_lock_to_gpu(
    kv_cache: "_KVCache", tasks: Sequence[BatchedLockTarget]
) -> list["_SharedPageLock"]:
    "Lock pages after migrating all pages to GPU. If migration fails, no locking happens."
    storage = kv_cache.manager._storage
    assert not tasks or storage is get_uniform_attribute(tasks, lambda p: p.page.manager)
    requirements = filled_list(0, storage.num_pool_groups)
    scheduled_for_eviction = [t.page.scheduled_for_eviction for t in tasks]
    for t, e in zip(tasks, scheduled_for_eviction):
        if e:
            storage.exclude_from_eviction(t.page)
        if t.page.cache_level == GPU_LEVEL:
            continue
        requirements[storage.get_pool_group_index(t.life_cycle)] += 1

    try:
        storage.prepare_free_slots(GPU_LEVEL, requirements)
        partitioned = partition(
            tasks, lambda p: (p.page.cache_level, storage.get_pool_group_index(p.life_cycle))
        )
        for (lvl, pg_idx), part in partitioned.items():
            if lvl == GPU_LEVEL:
                continue
            storage._batched_migrate(
                pg_idx, GPU_LEVEL, lvl, [p.page for p in part], update_src=True
            )
    except Exception:
        for t, e in zip(tasks, scheduled_for_eviction):
            if e:
                storage.schedule_for_eviction(t.page)
        raise
    stream_wait_events(kv_cache.cuda_stream, (p.page.ready_event for p in tasks))
    return [
        page.lock(kv_cache, beam_index, ordinal, life_cycle, skip_wait=True)
        for page, beam_index, ordinal, life_cycle in tasks
    ]

File: cpp/tensorrt_llm/batch_manager/kv_cache_manager_v2/page.cpp
/*
 * SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
 * SPDX-License-Identifier: Apache-2.0
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "kv_cache_manager_v2/page.h"
#include "kv_cache_manager_v2/common.h"
#include "kv_cache_manager_v2/exceptions.h"
#include "kv_cache_manager_v2/kvCache.h"        // for KvCache
#include "kv_cache_manager_v2/storageManager.h" // for StorageManager

#include "tensorrt_llm/common/assert.h"
#include <stdexcept>

namespace tensorrt_llm::batch_manager::kv_cache_manager_v2
{

// ---------------------------------------------------------------------------
// Page
// ---------------------------------------------------------------------------

Page::Page(StorageManager* mgr, LifeCycleId lc, CacheLevel level, Priority prio)
    : manager(mgr)
    , lifeCycle(lc)
    , cacheLevel(level)
    , priority(prio)
    , nodeRef(std::nullopt)
{
}

Page::~Page()
{
    TLLM_CHECK_DEBUG_WITH_INFO(status() == PageStatus::DROPPABLE && !scheduledForEviction(),
        "Page destroyed while still held or scheduled for eviction");
    if (hasValidSlot())
    {
        Slot s;
        s.setSlotId(slotId());
        s.readyEvent = std::move(readyEvent);
        resetSlot();
        manager->releaseSlot(lifeCycle, cacheLevel, std::move(s));
    }
}

PageStatus Page::status() const noexcept
{
    auto h = holder.lock();
    if (!h)
        return PageStatus::DROPPABLE;
    if (h->uniqLock.expired())
        return PageStatus::HELD;
    return PageStatus::LOCKED;
}

SharedPtr<PageHolder> Page::hold()
{
    // Return existing holder if any.
    auto h = holder.lock();
    if (h)
        return h;

    auto self = sharedFromThis();
    h = makeShared<PageHolder>(self);
    holder = h;

    // If we were scheduled for eviction but are no longer evictable (just got held), remove.
    if (scheduledForEviction())
    {
        if (!manager->isEvictable(*this))
        {
            manager->excludeFromEviction(*this);
            TLLM_CHECK_DEBUG(!scheduledForEviction());
        }
    }
    return h;
}

SharedPageLock Page::lock(KvCache& kvCache, BeamIndex beamIndex, BlockOrdinal ordinal, LifeCycleId lc, bool skipWait)
{
    return hold()->lock(kvCache, beamIndex, ordinal, lc, skipWait);
}

// ---------------------------------------------------------------------------
// CommittedPage
// ---------------------------------------------------------------------------

CommittedPage::CommittedPage(StorageManager* mgr, SharedPtr<Block> blk, LifeCycleId lc, CacheLevel level, Priority prio)
    : Page(mgr, lc, level, prio)
    , block(blk.get())
{
}

SsmCommittedPage::SsmCommittedPage(
    StorageManager* mgr, SharedPtr<Block> blk, LifeCycleId lc, CacheLevel level, Priority prio, int numTokensInBlock_)
    : CommittedPage(mgr, std::move(blk), lc, level, prio)
    , numTokensInBlock(numTokensInBlock_)
{
    TLLM_CHECK_DEBUG(numTokensInBlock_ > 0);
}

CommittedPage::~CommittedPage()
{
    if (block != nullptr)
    {
        // unlinkPage nulls storage[lc]->block (i.e. our member), so capture
        // the block pointer first. Pass `this` as the expected page: if a newer
        // page already replaced us in the slot (e.g. a larger SSM snapshot), the
        // slot is left alone and prev is nullptr — skip stale-block cleanup then.
        Block* blk = block;
        auto* prev = blk->unlinkPage(lifeCycle, this);
        if (prev != nullptr)
        {
            TLLM_CHECK_DEBUG_WITH_INFO(prev == this, "unlinkPage returned unexpected page");
            LifeCycle const& lc = manager->lifeCycles().getLifeCycle(lifeCycle);
            Block::clearStaleBlocksAfterPageUnlink(*blk, lifeCycle, lc);
        }
    }
    // Delegate slot release to Page::~Page().
}

// ---------------------------------------------------------------------------
// UncommittedPage
// ---------------------------------------------------------------------------

UncommittedPage::UncommittedPage(KvCache& kvc, BlockOrdinal ord, LifeCycleId lc, CacheLevel level, BeamIndex bi)
    : Page(kvc.storageManager(), lc, level, kvc.getPriority(ord, lc))
    , kvCache(&kvc)
    , ordinal(ord)
    , beamIndex(bi)
{
}

UncommittedPage::~UncommittedPage()
{
    // Mirrors Python UncommittedPage.__del__: for attention LCs, the page must be either:
    //  - part of an SSM lifecycle (different rules),
    //  - at an ordinal beyond the current block list (block already removed),
    //  - the slot at this position is null, CommittedPage, or this page itself (self-destruction).
    // The "p == this" condition is C++-specific: std::variant destroys the old value before
    // switching to monostate, so during destruction the slot still references this page.
    if (TLLM_UNLIKELY(gDebug))
    {
        auto ssmLcId = manager->lifeCycles().ssmLifeCycleId();
        bool isSsm = ssmLcId.has_value() && lifeCycle == *ssmLcId;
        if (!isSsm)
        {
            [[maybe_unused]] bool blockRemoved = kvCache->blocks().size() <= ordinal;
            [[maybe_unused]] bool pageOk = true;
            if (!blockRemoved)
            {
                auto const& bp = kvCache->blocks()[ordinal].pages[beamIndex][lifeCycle];
                auto page = blockPageGetPage(bp);
                pageOk
                    = blockPageIsNull(bp) || page.get() == this || dynamicPointerCast<CommittedPage>(page) != nullptr;
            }
            TLLM_CHECK_WITH_INFO(
                blockRemoved || pageOk, "UncommittedPage destroyed but slot still holds a different uncommitted page");
        }
    }
    // Delegate slot release to Page::~Page().
}

SharedPtr<CommittedPage> UncommittedPage::convertToCommitted(SharedPtr<Block> blk, CachedCudaEvent readyEv)
{
    TLLM_CHECK_DEBUG(!scheduledForEviction());
    TLLM_CHECK_DEBUG_WITH_INFO(
        blk->storage.at(lifeCycle) == nullptr, "Block slot for this lifecycle already has a committed page");
    TLLM_CHECK_DEBUG_WITH_INFO(status() == PageStatus::DROPPABLE, "Release holder/lock before converting");

    // Set the ready event before transfer (matches Python: self.ready_event = ready_event).
    this->readyEvent = std::move(readyEv);

    auto committed = makeShared<CommittedPage>(manager, blk, lifeCycle, cacheLevel, priority);
    // Move slot id to the committed page; invalidate our slot.
    committed->setSlotId(slotId()); // asserts valid
    committed->readyEvent = std::move(readyEvent);
    resetSlot();
    readyEvent = CachedCudaEvent::makeNull();

    TLLM_CHECK_DEBUG(!hasValidSlot() && readyEvent.isClosed());
    TLLM_CHECK_DEBUG_WITH_INFO(committed->hasValidSlot(), "committed page must have a valid slot after transfer");

    // Register in block storage.
    blk->storage.at(lifeCycle) = committed.get();

    return committed;
}

SharedPtr<SsmCommittedPage> UncommittedPage::convertToSsmCommitted(
    SharedPtr<Block> blk, CachedCudaEvent readyEv, int numTokensInBlock)
{
    TLLM_CHECK_DEBUG(!scheduledForEviction());
    TLLM_CHECK_DEBUG_WITH_INFO(
        blk->storage.at(lifeCycle) == nullptr, "Block slot for this lifecycle already has a committed page");
    TLLM_CHECK_DEBUG_WITH_INFO(status() == PageStatus::DROPPABLE, "Release holder/lock before converting");

    this->readyEvent = std::move(readyEv);

    auto committed = makeShared<SsmCommittedPage>(manager, blk, lifeCycle, cacheLevel, priority, numTokensInBlock);
    committed->setSlotId(slotId()); // asserts valid
    committed->readyEvent = std::move(readyEvent);
    resetSlot();
    readyEvent = CachedCudaEvent::makeNull();

    TLLM_CHECK_DEBUG(!hasValidSlot() && readyEvent.isClosed());
    TLLM_CHECK_DEBUG_WITH_INFO(committed->hasValidSlot(), "committed page must have a valid slot after transfer");

    blk->storage.at(lifeCycle) = committed.get();

    return committed;
}

// ---------------------------------------------------------------------------
// PageHolder
// ---------------------------------------------------------------------------

PageHolder::PageHolder(SharedPtr<Page> p)
    : page(std::move(p))
{
}

PageHolder::~PageHolder()
{
    TLLM_CHECK_DEBUG_WITH_INFO(uniqLock.expired(), "PageHolder destroyed while lock still active");

    page->holder.reset(); // clear back-reference
    auto const manager = page->manager;

    // If it's a committed page, schedule for eviction (if evictable).
    if (page->isCommitted())
    {
        if (!page->scheduledForEviction())
            manager->scheduleForEviction(*page);

        // If the block is orphan, exclude from eviction immediately.
        auto* cp = dynamic_cast<CommittedPage*>(page.get());
        if (cp)
        {
            if (cp->block == nullptr || cp->block->isOrphan())
                manager->excludeFromEviction(*page);
        }
    }
    else
    {
        // Uncommitted page: if scheduled for eviction, remove it.
        if (page->scheduledForEviction())
            manager->excludeFromEviction(*page);
    }
}

SharedPageLock PageHolder::lock(
    KvCache& kvCache, BeamIndex beamIndex, BlockOrdinal ordinal, LifeCycleId lc, bool skipWait)
{
    // Create or reuse UniqPageLock.
    auto ul = uniqLock.lock();
    if (!ul)
    {
        ul = makeShared<UniqPageLock>(sharedFromThis());
        uniqLock = ul;
    }

    // Remove from eviction queue if scheduled.
    if (page->scheduledForEviction())
    {
        page->manager->excludeFromEviction(*page);
        TLLM_CHECK_DEBUG(!page->scheduledForEviction());
    }

    return ul->share(kvCache, beamIndex, ordinal, lc, skipWait);
}

// ---------------------------------------------------------------------------
// UniqPageLock
// ---------------------------------------------------------------------------

UniqPageLock::UniqPageLock(SharedPtr<PageHolder> h)
    : holder(std::move(h))
{
    if (holder->page->cacheLevel != kGpuLevel)
        throw LogicError("Lock can only be applied to GPU-memory pages");
}

UniqPageLock::~UniqPageLock()
{
    Page& p = *page();
    TLLM_CHECK_DEBUG(p.cacheLevel == kGpuLevel && !p.scheduledForEviction());
    // Set readyEvent to the merged finish events of all readers. For committed (read-only)
    // pages, this means the next reader will wait for prior reads to complete, which is
    // unnecessary but correct. See the CommittedPage comment in page.h for rationale.
    p.readyEvent = mergeEvents(finishEvents);

    // Clear the holder's lock reference.
    TLLM_CHECK_DEBUG(holder);
    holder->uniqLock.reset();

    // Optimized path (mirrors Python): set holder=nullptr, then check if still evictable.
    auto holderCopy = std::move(holder);
    holder = nullptr;

    // If the page is not droppable (still held by someone else) and evictable,
    // schedule for eviction.
    if (p.status() != PageStatus::DROPPABLE)
    {
        auto const manager = p.manager;
        if (manager->isEvictable(p))
            manager->scheduleForEviction(p);
    }
}

void UniqPageLock::notifyFinish(CachedCudaEvent event)
{
    finishEvents.push_back(std::move(event));
    // Avoid unbounded growth for system prompt pages shared by all requests.
    if (finishEvents.size() > 32)
    {
        CachedCudaEvent merged = mergeEvents(finishEvents);
        finishEvents.clear();
        finishEvents.push_back(std::move(merged));
    }
}

SharedPtr<Page> const& UniqPageLock::page() const
{
    TLLM_CHECK_DEBUG(holder && holder->page);
    return holder->page;
}

SharedPageLock UniqPageLock::share(
    KvCache& kvCache, BeamIndex beamIndex, BlockOrdinal ordinal, LifeCycleId lc, bool skipWait)
{
    return SharedPageLock(sharedFromThis(), kvCache, beamIndex, ordinal, lc, skipWait);
}

// ---------------------------------------------------------------------------
// SharedPageLock
// ---------------------------------------------------------------------------

SharedPageLock::SharedPageLock(SharedPtr<UniqPageLock> ul, KvCache& kvCache, BeamIndex beamIndex, BlockOrdinal ordinal,
    LifeCycleId lc, bool skipWait)
    : mUniqLock(std::move(ul))
    , mUser{&kvCache, beamIndex, ordinal, lc}
{
    if (!skipWait)
        page()->readyEvent.waitInStream(reinterpret_cast<CudaStream>(kvCache.cudaStream()));

    acquirePageIndex();
}

SharedPageLock::~SharedPageLock()
{
    if (mUniqLock)
        unlock();
}

SharedPageLock::SharedPageLock(SharedPageLock&& other) noexcept
    : mUniqLock(std::move(other.mUniqLock))
    , mUser(std::move(other.mUser))
{
}

SharedPageLock& SharedPageLock::operator=(SharedPageLock&& other) noexcept
{
    if (this != &other)
    {
        if (mUniqLock)
            unlock();
        mUniqLock = std::move(other.mUniqLock);
        mUser = std::move(other.mUser);
    }
    return *this;
}

SharedPtr<Page> const& SharedPageLock::page() const
{
    TLLM_CHECK_DEBUG(mUniqLock);
    return mUniqLock->page();
}

SharedPtr<Page> SharedPageLock::unlock()
{
    TLLM_CHECK_DEBUG(mUniqLock);

    // Record finish event from the KvCache stream.
    mUniqLock->notifyFinish(mUser.kvCache->finishEvent());

    releasePageIndex();
    auto p = page(); // copy shared_ptr before reset
    mUniqLock.reset();
    return p;
}

void SharedPageLock::acquirePageIndex()
{
    auto* kvc = mUser.kvCache;
    auto& pg = *page();
    int old = kvc->updateBasePageIndex(
        mUser.beamIndex, mUser.ordinal, mUser.lifeCycle, slotIdToPageIndexValue(pg.slotId()));
    // Mirrors Python assertion: old base index must be BAD (prevents double-locking same slot).
    TLLM_CHECK_DEBUG_WITH_INFO(
        old == kBadPageIndex.value(), "Double-lock: page index already acquired for this (beam, ordinal, lc)");
    (void) old;
}

void SharedPageLock::releasePageIndex()
{
    int oldBaseIndex
        = mUser.kvCache->updateBasePageIndex(mUser.beamIndex, mUser.ordinal, mUser.lifeCycle, kBadPageIndex.value());
    // Mirrors Python assertion: old base index must match this page's slot ID.
    TLLM_CHECK_DEBUG(oldBaseIndex == slotIdToPageIndexValue(page()->slotId()));
    (void) oldBaseIndex;
}

// ---------------------------------------------------------------------------
// batchedLockToGpu
// ---------------------------------------------------------------------------

std::vector<SharedPageLock> batchedLockToGpu(KvCache& kvCache, std::vector<BatchedLockTarget> const& targets)
{
    auto* storeMgr = kvCache.storageManager();
    TLLM_CHECK_DEBUG(storeMgr);
    // All pages must belong to the same storage manager.
    TLLM_CHECK_DEBUG(targets.empty()
        || std::all_of(targets.begin(), targets.end(), [&](auto const& t) { return t.page->manager == storeMgr; }));

    // Determine how many GPU slots are needed per pool group.
    TypedVec<PoolGroupIndex, SlotCount> requirements(storeMgr->numPoolGroups(), 0);
    std::vector<bool> wasScheduled(targets.size(), false);

    for (size_t i = 0; i < targets.size(); ++i)
    {
        auto const& t = targets[i];
        wasScheduled[i] = t.page->scheduledForEviction();
        if (wasScheduled[i])
            storeMgr->excludeFromEviction(*t.page);
        if (t.page->cacheLevel != kGpuLevel)
        {
            PoolGroupIndex pgIdx = storeMgr->getPoolGroupIndex(t.lifeCycle);
            requirements[pgIdx] += 1;
        }
    }

    try
    {
        MigrationRecorder const migrationRecorder
            = [&kvCache](std::vector<SharedPtr<Page>> const& pages, std::vector<Slot> const& slots, CacheLevel srcLevel,
                  CacheLevel dstLevel) { kvCache._recordMigratedSlots(pages, slots, srcLevel, dstLevel); };
        DropRecorder const dropRecorder = [&kvCache](std::vector<SharedPtr<Page>> const& pages, CacheLevel cacheLevel)
        { kvCache._recordDroppedPages(pages, cacheLevel); };
        storeMgr->prepareFreeSlots(kGpuLevel, requirements, migrationRecorder, dropRecorder);
        // Migrate non-GPU pages.
        storeMgr->batchedMigrateToGpu(targets, kvCache, migrationRecorder);
    }
    catch (...)
    {
        // Restore eviction scheduling.
        for (size_t i = 0; i < targets.size(); ++i)
            if (wasScheduled[i])
                storeMgr->scheduleForEviction(*targets[i].page);
        throw;
    }

    // Wait for all ready events on KvCache's stream (deduplicated).
    {
        std::vector<CachedCudaEvent const*> readyEvents;
        readyEvents.reserve(targets.size());
        for (auto const& t : targets)
            readyEvents.push_back(&t.page->readyEvent);
        streamWaitEvents(reinterpret_cast<CudaStream>(kvCache.cudaStream()), readyEvents);
    }

    // Lock all pages.
    std::vector<SharedPageLock> locks;
    locks.reserve(targets.size());
    for (auto const& t : targets)
        locks.emplace_back(t.page->lock(kvCache, t.beamIndex, t.ordinal, t.lifeCycle,
            /*skipWait=*/true));
    return locks;
}

// ---------------------------------------------------------------------------
// ScratchSlotLock
// ---------------------------------------------------------------------------

ScratchSlotLock::ScratchSlotLock(Slot slot, KvCache& owner, LifeCycleId lifeCycle, bool skipWait)
    : mOwner(&owner)
    , mLifeCycle(lifeCycle)
{
    if (!skipWait)
    {
        slot.readyEvent.waitInStream(reinterpret_cast<CudaStream>(owner.cudaStream()));
    }
    mSlot.setSlot(slot);
}

ScratchSlotLock::~ScratchSlotLock()
{
    if (mSlot.hasValidSlot())
    {
        try
        {
            unlock();
        }
        catch (...)
        {
        }
    }
}

ScratchSlotLock::ScratchSlotLock(ScratchSlotLock&& other) noexcept
    : mSlot(std::move(other.mSlot))
    , mOwner(other.mOwner)
    , mLifeCycle(other.mLifeCycle)
{
    // Invalidate moved-from: mSlot move only transfers readyEvent, not slotId (trivially-copyable).
    other.mSlot.resetSlot();
    other.mOwner = nullptr;
}

ScratchSlotLock& ScratchSlotLock::operator=(ScratchSlotLock&& other) noexcept
{
    if (this != &other)
    {
        if (mSlot.hasValidSlot())
        {
            try
            {
                unlock();
            }
            catch (...)
            {
            }
        }
        mSlot = std::move(other.mSlot);
        other.mSlot.resetSlot(); // Invalidate moved-from slotId.
        mOwner = other.mOwner;
        mLifeCycle = other.mLifeCycle;
        other.mOwner = nullptr;
    }
    return *this;
}

Slot ScratchSlotLock::detachSlot()
{
    TLLM_CHECK_DEBUG(mSlot.hasValidSlot());
    Slot result;
    result.setSlot(mSlot);
    return result;
}

void ScratchSlotLock::unlock()
{
    TLLM_CHECK_DEBUG(mSlot.hasValidSlot());
    mSlot.readyEvent = mOwner->finishEvent();
    mOwner->storageManager()->releaseSlot(mLifeCycle, kGpuLevel, std::move(mSlot));
    TLLM_CHECK_DEBUG(!mSlot.hasValidSlot());
}

} // namespace tensorrt_llm::batch_manager::kv_cache_manager_v2
